File: leetcode_python/3.Longest_Substring_Without_Repeating_Characters.py
#!/usr/bin/env python3
# coding: utf-8

# Time complexity: O()
# Space complexity: O()

# https://leetcode.com/problems/longest-substring-without-repeating-characters/submissions/


# similar to 340

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        d = {}
        max_len = 0
        for i, val in enumerate(s):
            if val in d and d[val] >= left:
                left = d[val] + 1
            d[val] = i
            max_len = max(max_len, i - left + 1)
        return max_len

File: leetcode_java/3.Longest_Substring_Without_Repeating_Characters.java
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

class Solution {
  public int lengthOfLongestSubstring(String s) {
    HashMap<Character, Integer> map = new HashMap<>();

    int i = 0;
    int res = 0;

    for (int j = 0; j < s.length(); j++) {
      if (map.containsKey(s.charAt(j))) {
        i = Math.max(i, map.get(s.charAt(j)) + 1);
      }
      map.put(s.charAt(j), j);
      res = Math.max(res, j - i + 1);
    }

    return res;
  }
}